File: python/210.py
import collections
xrange = range


class Solution(object):
    def dfs(self, pre, i, finished, saw, order):
        if i in finished:
            return True
        if i in saw:
            return False
        saw.add(i)
        res = True
        for j in pre[i]:
            res = res and self.dfs(pre, j, finished, saw, order)
        finished.add(i)
        order.append(i)
        return res

    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        pre = collections.defaultdict(set)
        for u, v in prerequisites:
            pre[u].add(v)
        finished = set()
        saw = set()
        order = []
        for i in xrange(numCourses):
            if not self.dfs(pre, i, finished, saw, order):
                return []
        return order

File: cpp/210.cpp
#include <set>
#include <stack>
#include <vector>
#include <utility>
#include <iostream>
using namespace std;

class Solution {
public:
  vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
    vector<vector<int>> pre(numCourses);
    for(auto &p : prerequisites){
      int i = p.first, j = p.second;
      pre[i].push_back(j);
    }
    bool finished[numCourses], saw[numCourses];
    vector<int> order;

    for(int i=0; i<numCourses; i++){
      if(pre[i].size() == 0){
        finished[i] = saw[i] = true;
        order.push_back(i);
      } else
        finished[i] = saw[i] = false;
    }

    stack<int> s;

    for(int i=0; i<numCourses; i++){
      if(finished[i]) continue;
      s.push(i);
      while(!s.empty()){
        int u = s.top();
        saw[u] = true;
        bool done = true;
        for(int v : pre[u])
          if(!finished[v]){
            if(saw[v]) return {};
            done = false;
            s.push(v);
          }
        if(done){
          finished[u] = true; s.pop();
          order.push_back(u);
        }
      }
    }
    return order;
  }
};
